android/src/main/java/com/magnet/mmx/client/api/MMXMessage.java

```diff
@@ -832,4 +889,71 @@ static MMXMessage getSendingMessage(String messageId) {
       return listenerPair.message;
     }
   }
+
+  private void uploadAttachments(MMXPayload payload, final String messageId, final Attachment.AttachmentTrasferLister lister) {
+    for(final Attachment attachment : mAttachments) {
+      if(Attachment.Status.COMPLETE == attachment.getStatus()
+          && StringUtil.isNotEmpty(attachment.getAttachmentId())) {
+        if(null != lister) {
+          lister.onComplete(attachment);
+        }
+        continue;
+      }
+
+      if(null != lister) {
+        lister.onStart(attachment);
+      }
+
+      final CountDownLatch uploadSignal = new CountDownLatch(1);
+      attachment.upload(new Attachment.AttachmentTrasferLister() {
+        @Override public void onStart(Attachment attachment) {
+          if(null != lister) {
+            lister.onStart(attachment);
+          }
+        }
+
+        @Override public void onProgress(Attachment attachment, long l) {
+          if(null != lister) {
+            lister.onProgress(attachment, l);
+          }
+        }
+
+        @Override public void onComplete(Attachment attachment) {
+          if(null != lister) {
+            lister.onComplete(attachment);
+          }
+          uploadSignal.countDown();
+        }
+
+        @Override public void onError(Attachment attachment, Throwable throwable) {
+          if(null != lister) {
+            lister.onError(attachment, throwable);
+          }
+          uploadSignal.countDown();
+        }
+      });
+
+      try {
+        uploadSignal.await(60, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+
+      if(Attachment.Status.COMPLETE == attachment.getStatus()
+          && StringUtil.isNotEmpty(attachment.getAttachmentId())) {
+        lister.onComplete(attachment);
+        Log.d(TAG, "Attachment " + attachment.getName() + " is uploaded successfully.");
+      } else {
+        String message = "Failed to upload attachment " + attachment.getName();
+        Log.d(TAG, message);
+        if(uploadSignal.getCount() > 0) {
+          if(null != lister) {
+            lister.onError(attachment, new Exception(message));
+          }
+        }
+      }
+    }
+
+    payload.setMetaData(CONTENT_ATTACHMENTS, GsonData.getGson().toJson(mAttachments));
```

Comment:
Use setMmxMetaData() instead of setMetaData().  The mmxMetaData is reserved for our SDK.  The metaData is for developers.